eax/eax.go

```diff
@@ -0,0 +1,182 @@
+// Copyright (C) 2019 ProtonTech AG
+
+// Package eax provides an implementation of the EAX
+// (encrypt-authenticate-translate) mode of operation, as described in
+// Bellare, Rogaway, and Wagner "THE EAX MODE OF OPERATION: A TWO-PASS
+// AUTHENTICATED-ENCRYPTION SCHEME OPTIMIZED FOR SIMPLICITY AND EFFICIENCY."
+// In FSE'04, volume 3017 of LNCS, 2004
+
+package eax
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"errors"
+	"golang.org/x/crypto/internal/byteutil"
+)
+
+const (
+	defaultTagSize   = 16
+	defaultNonceSize = 16
+)
+
+type eax struct {
+	block     cipher.Block // Only AES-{128, 192, 256} supported
+	tagSize   int          // At least 12 bytes recommended
+	nonceSize int
+}
+
+//
+// AEAD interface implementation
+//
+
+func (e *eax) NonceSize() int {
+	return e.nonceSize
+}
+
+func (e *eax) Overhead() int {
+	return e.tagSize
+}
+
+func (e *eax) Seal(dst, nonce, plaintext, data []byte) []byte {
+	return e.Encrypt(nonce, plaintext, data)
+}
+
+func (e* eax) Open(dst, nonce, ciphertext, data []byte) ([]byte, error) {
+	return e.Decrypt(nonce, ciphertext, data)
+}
+
+// NewEAX returns an EAX instance with AES-{keyLength} and default tag and
+// nonce lengths.
+func NewEAX(key []byte) (cipher.AEAD, error) {
+	return NewEAXWithNonceAndTagSize(key, defaultNonceSize, defaultTagSize)
+}
+
+// NewEAXWithNonceAndTagSize returns an EAX instance with AES-{keyLength} and
+// given nonce and tag lengths in bytes. Panics on zero nonceSize and
+// exceedingly long tags.
+//
+// It is recommended to use at least 12 bytes as tag length (see, for instance,
+// NIST SP 800-38D).
+//
+// Only to be used for compatibility with existing cryptosystems with
+// non-standard parameters. For all other cases, prefer NewEAX.
+func NewEAXWithNonceAndTagSize(
+	key []byte, nonceSize, tagSize int) (cipher.AEAD, error) {
+	if nonceSize == 0 {
+		return nil, errors.New("Can't initialize EAX instance with nonceSize = 0")
+	}
+	aesCipher, err := aes.NewCipher(key)
+	if err != nil {
+		panic(err)
+	}
+	if tagSize > aesCipher.BlockSize() {
+		return nil, errors.New("Custom tag length exceeds blocksize")
+	}
+	return &eax{
+		block:     aesCipher,
+		tagSize:   tagSize,
+		nonceSize: nonceSize,
+	}, nil
+}
+
+func (e *eax) TagSize() int {
+	return e.tagSize
+}
+
+// Encrypt function. Returns a byte array containing the concatenation
+// of the ciphertext and the validation tag.
+func (e *eax) Encrypt(nonce, plaintext, adata []byte) []byte {
+
+	omacNonce := e.omacT(0, nonce)
```

Comment:
It might make sense to have another check on the nonce length here (and in `Decrypt`). The GCM implementation checks whether it's `e.nonceLength`, but the EAX paper explicitly states

> it might be convenient, for example, to allow nonces that are initially 1 byte, and then grow to 2 bytes, and so forth

so there's no requirement there to specify the nonce size in advance, and it can differ per message. Nevertheless, I think GCM is the same (but haven't checked), so maybe it makes sense to have that extra restriction here too. (Haven't checked OCB either.)

Reply:
Good idea, it looks better to check that the given length does not exceed the one specified in the EAX instance, and same for OCB. Actually, growing nonces are currently implemented (they are padded) and covered by tests, so I'm adding 
```
	if len(nonce) > e.nonceSize {
		return nil, errors.New("EAX: Incorrect nonce size")
	}
```

Reply:
However, the AEAD interface does not specify an error return for the `Seal` function (i.e., it always returns `[]byte` as opposed to `([]byte, error)` so this check can only be performed in `Open`. 
Alternatively, one could do the test in `Open` and panic at failure, or modify the AEAD interface (involves changing GCM too...)

Reply:
Using panic too, as you pointed out this is the behaviour in [gcm.go](https://golang.org/src/crypto/cipher/gcm.go)